File: 강의 자료/02-알고리즘/13-위상정렬/문제별 코드/1005-ACM Craft/solution.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N, M;
    static int[] indeg, T_done, T;
    static ArrayList<Integer>[] adj;

    static void input() {
        N = scan.nextInt();
        M = scan.nextInt();
        adj = new ArrayList[N + 1];
        indeg = new int[N + 1];
        T = new int[N + 1];
        T_done = new int[N + 1];
        for (int i = 1; i <= N; i++) {
            adj[i] = new ArrayList<>();
            T[i] = scan.nextInt();
        }
        for (int i = 0; i < M; i++) {
            int x = scan.nextInt(), y = scan.nextInt();
            adj[x].add(y);
            // indegree 계산하기
            indeg[y]++;
        }
    }

    static void pro() {
        Deque<Integer> queue = new LinkedList<>();
        // 제일 앞에 "정렬될 수 있는" 정점 찾기
        for (int i = 1; i <= N; i++)
            if (indeg[i] == 0) {
                queue.add(i);
                T_done[i] = T[i];
            }

        // 위상 정렬 순서대로 T_done 계산을 함께 해주기
        while (!queue.isEmpty()) {
            int x = queue.poll();
            for (int y : adj[x]) {
                indeg[y]--;
                if (indeg[y] == 0) queue.add(y);
                T_done[y] = Math.max(T_done[y], T_done[x] + T[y]);
            }
        }
        int W = scan.nextInt();
        System.out.println(T_done[W]);
    }

    public static void main(String[] args) {
        int Q = scan.nextInt();
        while (Q > 0) {
            Q--;
            input();
            pro();
        }
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: 강의 자료/02-알고리즘/13-위상정렬/문제별 코드/1005-ACM Craft/solution.py
import sys
from collections import deque
si = sys.stdin.readline

Q = int(si())
for _ in range(Q):
    n, m = map(int, si().split())
    T = [0] + list(map(int, si().split()))
    T_done = [0] * (n + 1)
    con = [[] for _ in range(n + 1)]
    indeg = [0] * (n + 1)

    for _ in range(m):
        x, y = map(int, si().split())
        con[x].append(y)
        # indegree 계산하기
        indeg[y] += 1

    # 제일 앞에 "정렬될 수 있는" 정점 찾기
    Q = deque()
    for i in range(1, n + 1):
        if indeg[i] == 0:
            Q.append(i)
            T_done[i] = T[i]

    # 위상 정렬 순서대로 T_done 계산을 함께 해주기
    while Q:
        x = Q.popleft()
        for y in con[x]:
            indeg[y] -= 1
            if indeg[y] == 0:
                Q.append(y)
            T_done[y] = max(T_done[y], T_done[x] + T[y])
    
    W = int(si())
    print(T_done[W])
